File: core/src/main/java/com/starshootercity/abilities/LaunchIntoAir.java
package com.starshootercity.abilities;

import com.starshootercity.OriginSwapper;
import com.starshootercity.cooldowns.CooldownAbility;
import com.starshootercity.cooldowns.Cooldowns;
import net.kyori.adventure.key.Key;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerToggleSneakEvent;
import org.bukkit.util.Vector;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class LaunchIntoAir implements VisibleAbility, Listener, CooldownAbility {
    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:launch_into_air");
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getDescription() {
        return OriginSwapper.LineData.makeLineFor("Every 30 seconds, you are able to launch about 20 blocks up into the air.", OriginSwapper.LineData.LineComponent.LineType.DESCRIPTION);
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getTitle() {
        return OriginSwapper.LineData.makeLineFor("Gift of the Winds", OriginSwapper.LineData.LineComponent.LineType.TITLE);
    }

    @EventHandler
    public void onSneakToggle(PlayerToggleSneakEvent event) {
        if (!event.isSneaking()) return;
        AbilityRegister.runForAbility(event.getPlayer(), getKey(), () -> {
            if (event.getPlayer().isGliding()) {
                if (hasCooldown(event.getPlayer())) return;
                setCooldown(event.getPlayer());
                event.getPlayer().setVelocity(event.getPlayer().getVelocity().add(new Vector(0, 2, 0)));
            }
        });
    }

    @Override
    public Cooldowns.CooldownInfo getCooldownInfo() {
        return new Cooldowns.CooldownInfo(600, "launch");
    }
}

File: core/src/main/java/com/starshootercity/abilities/LaunchIntoAir.kt
package com.starshootercity.abilities

import com.starshootercity.OriginSwapper.LineData.Companion.makeLineFor
import com.starshootercity.OriginSwapper.LineData.LineComponent
import com.starshootercity.cooldowns.CooldownAbility
import com.starshootercity.cooldowns.Cooldowns.CooldownInfo
import net.kyori.adventure.key.Key
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerToggleSneakEvent
import org.bukkit.util.Vector

class LaunchIntoAir : VisibleAbility, Listener, CooldownAbility {

    override fun getKey(): Key {
        return Key.key("origins:launch_into_air")
    }

    override val description: MutableList<LineComponent?> = makeLineFor(
        "Every 30 seconds, you are able to launch about 20 blocks up into the air.",
        LineComponent.LineType.DESCRIPTION
    )

    override val title: MutableList<LineComponent?> = makeLineFor(
        "Gift of the Winds",
        LineComponent.LineType.TITLE
    )

    @EventHandler
    fun onSneakToggle(event: PlayerToggleSneakEvent) {
        if (!event.isSneaking) return
        val player = event.player
        runForAbility(player) { p ->
            if (p.isGliding && !hasCooldown(p)) {
                setCooldown(p)
                p.velocity = p.velocity.add(Vector(0, 2, 0))
            }
        }
    }

    override val cooldownInfo: CooldownInfo
        get() = CooldownInfo(600, "launch")
}
